### PR TITLE
Implement the Cargo half of pipelined compilation

### PR DESCRIPTION
**Pull request by alexcrichton**
This implements what I hope is 90% ish of the Cargo side of the work of #6660. The goal here is to eventually execute rlib compilations in a more "pipelined fashion" which involves taking more advantage of the parallelism that most machines have nowadays.

There is a lot of comments and wording on each commit and in documentation, so I won't repeat it all here too much!

We probably don't want to land this too eagerly, but it should in theory pass all tests and be ready to land at any time (just not bring any benefit yet). I'd be fine holding off until the rustc side of pipelining is implemented and then we can polish it off here and get this ready to go!

### CODE REVIEW

**Comment by alexcrichton** [comment]
cc @nnethercote, you're likely very interested in this!

r? @ehuss you're probably the best to take a look at it

@ehuss I'm not 100% satisfied with how management of pipelined compilation ended up, but overall I think it turned out better than expected. If you've got ideas of how to better model this in Cargo please let me know!

File: src/cargo/core/compiler/mod.rs [hunk 1]
```diff
@@ -787,7 +811,7 @@ fn build_base_args<'a, 'cfg>(
     if unit.mode.is_check() {
         cmd.arg("--emit=dep-info,metadata");
     } else {
-        cmd.arg("--emit=dep-info,link");
+        cmd.arg("--emit=dep-info,metadata,link");
```
Comment (ehuss) [inline src/cargo/core/compiler/mod.rs]: I'm a little concerned about unconditionally emitting metadata. I can't immediately think of a problem, so maybe it doesn't matter. The only thing I can think of is that binary-only crates emit an empty file.

File: src/cargo/core/compiler/build_config.rs [hunk 2]
```diff
@@ -127,6 +127,7 @@ pub enum CompileMode {
     Test,
     /// Building a target with `rustc` (lib or bin).
     Build,
+    BuildRmeta,
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply src/cargo/core/compiler/build_config.rs]: Oops, forgot to come back and knock this out

File: src/cargo/core/compiler/mod.rs (hunk 1 above)
Comment (alexcrichton) [reply src/cargo/core/compiler/mod.rs]: Ah this is one I meant to come back and address, it's actually quite easy to only `--emit=metadata` for `BuildRmeta` targets.

File: tests/testsuite/build.rs [hunk 3]
```diff
@@ -1925,6 +1921,9 @@ fn verbose_release_build_deps() {
             version = "0.0.0"
             authors = []
 
+            [lib]
+            crate-type = ['cdylib']
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply tests/testsuite/build.rs]: At the time I made this change it was a testament to my laziness about how to get the test assertions working again. With the current iteration, however, it's not necessary at all!

**Comment by alexcrichton** [comment]
> I might be a little uncomfortable landing this immediately without the rustc side being ready. I think the cleanup changes could be landed immediately, though. Would you want to post them in separate PRs? I don't have a sense of how long the rustc side will take.
> 
> I would also be a little uncomfortable landing this enabled by default. It seems like it might have a high risk of breaking some users.

I think I'd tend to agree as well. I don't mind splitting this up, but I think it may not be necessary in theory if we tun it off by default because I'm pretty confident at least that this is unlikely to regress anything out there in the wild if we default it to off.

Would you agree though? Don't mind splitting up, but wanted to double check :)

> Even though there are now extra .rmeta files, each .rlib file is overall smaller. Can you explain why that is?

Uh...

witchcraft?

Are you sure that you were using before/after cargo against the exact same compiler? Were before/after cargo the tip of this PR and the merge-base with master? 

I would also expect build directories to become much larger...


> There does seem to be a small performance penalty

I'm gonna try to dig into this, I'll get back to you.

> From a high level, I was wondering if you considered using multiple edges in the graph instead of multiple nodes? Something like this where the orange edges are "rmeta" edges:

I think that does conceptually make more sense yeah. I was talking with fitzgen over lunch and he recommended something similar where I think the translation is that instead of finishing a node in a dependency graph we finish an edge (or a set of edges). That would probably be easier to model than what is currently implemented and the job executor wouldn't need any change. I'll try to think more and see if it's reasonable to implement as part of this as well.

**Comment by nnethercote** [comment]
@ehuss https://github.com/rust-lang/rust/pull/60006 has my first attempt at the rustc side. It isn't right and more work is needed, but hopefully not too much more

@alexcrichton Is the code you've added assuming that the directive from rustc has a particular form?

**Comment by alexcrichton** [comment]
@nnethercote nah this isn't tied to anything, and it's effectively "mocked out" in the sense that all we need to do now is update a few lines in Cargo with an implementation of reading a signal from rustc and Cargo should be good to go!

**Comment by ehuss** [comment]
I'm fine with landing this if it is off by default.

> Are you sure that you were using before/after cargo against the exact same compiler? Were before/after cargo the tip of this PR and the merge-base with master?

I was using `CARGO_BUILD_PIPELINING=false` to switch.  Looking closer, the rlibs built with `--emit=dep-info,metadata,link` have only one CGU.  Does that disable multiple cgu's (perhaps via [`is_compatible_with_codegen_units_and_single_output_file`](https://github.com/rust-lang/rust/blob/c21fbfe7e310b9055ed6b7c46b7d37b831a516e3/src/librustc/session/config.rs#L154))?

**Comment by alexcrichton** [comment]
Nothing like trading one form of parallelism for another! That definitely sounds like a rustc bug and would also explain the samller build directory (wow I had no idea one CGU had that much file savings). I'll investigate that on the rustc side of things soon.

Once https://github.com/rust-lang/cargo/pull/6867 lands I'll rebase this and clean up the history a bit and then we can land with it by default turned off?

**Comment by alexcrichton** [comment]
> Does that disable multiple cgu's (perhaps via [`is_compatible_with_codegen_units_and_single_output_file`](https://github.com/rust-lang/rust/blob/c21fbfe7e310b9055ed6b7c46b7d37b831a516e3/src/librustc/session/config.rs#L154))?

You're [spot on](https://github.com/rust-lang/rust/pull/60270)! That would also explain the slower overall compile as well as smaller output artifacts for sure

**Comment by ehuss** [comment]
> Once #6867 lands I'll rebase this and clean up the history a bit and then we can land with it by default turned off?

If you don't want to try to use multiple edges instead of multiple nodes (which I think could be a little simpler code-wise, but not required), then that seems reasonable.

**Comment by alexcrichton** [comment]
Thinking about this I would like to test out that model and see how well it fits. It may be a bigger refactor but I suspect it's going to be more future proof. There's not a lot of time pressure right now since the rustc side of things still isn't ready yet, so no harm in exploring our options!

**Comment by alexcrichton** [comment]
Ok I implemented the edge-based strategy and wow is it so much nicer. There's basically no comparison between them, so I'm going to close this.

I'll open a new PR once I've got everything cleaned up and documented. Thanks again for the suggestion @ehuss :)

**Comment by alexcrichton** [comment]
An updated version (and much smaller) is posted at https://github.com/rust-lang/cargo/pull/6883

**Pull request closed without merging**